atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/queue/KvsSweepQueueProgress.java

```diff
@@ -0,0 +1,116 @@
+import com.palantir.atlasdb.schema.generated.TargetedSweepTableFactory;
+import com.palantir.util.PersistableBoolean;
+
+public class KvsSweepQueueProgress {
+    private static final TableReference TABLE_REF = TargetedSweepTableFactory.of()
+            .getSweepShardProgressTable(null).getTableRef();
+    private static final int NUMBER_OF_SHARDS_INDEX = 1024;
+    private static final long CAS_TIMESTAMP = 1L;
+
+    private final KeyValueService kvs;
+
+    public KvsSweepQueueProgress(KeyValueService kvs) {
+        this.kvs = kvs;
+    }
+
+    public long numberOfShards() {
+        return getOrInitializeTo(NUMBER_OF_SHARDS_INDEX, true, 1L);
+    }
+
+    public void updateNumberOfShards(int newNumber) {
+        Preconditions.checkArgument(newNumber <= 128);
+        increaseValue(NUMBER_OF_SHARDS_INDEX, true, newNumber);
+    }
+
+    public long lastSweptTimestamp(int shard, boolean conservative) {
+        return getOrInitializeTo(shard, conservative, 0L);
+    }
+
+    public void updateLastSweptTimestamp(int shard, boolean conservative, long timestamp) {
+        increaseValue(shard, conservative, timestamp);
+    }
+
+    private long getOrInitializeTo(int shard, boolean conservative, long initialValue) {
+        Map<Cell, Value> result = getEntry(shard, conservative);
+        if (result.isEmpty()) {
+            return initializeShard(shard, conservative, initialValue);
+        }
+        return getValue(result);
+    }
+
+    private Map<Cell, Value> getEntry(int shard, boolean conservative) {
+        return kvs.get(TABLE_REF, ImmutableMap.of(cellForShard(shard, conservative), CAS_TIMESTAMP));
+    }
+
+    private Cell cellForShard(int shard, boolean conservative) {
+        SweepShardProgressTable.SweepShardProgressRow row = SweepShardProgressTable.SweepShardProgressRow.of(shard,
+                PersistableBoolean.of(conservative).persistToBytes());
+        return Cell.create(row.persistToBytes(),
+                SweepShardProgressTable.SweepShardProgressNamedColumn.VALUE.getShortName());
+    }
+
+    private long getValue(Map<Cell, Value> entry) {
+        SweepShardProgressTable.Value value = SweepShardProgressTable.Value.BYTES_HYDRATOR.hydrateFromBytes(
+                Iterables.getOnlyElement(entry.values()).getContents());
+        return value.getValue();
+    }
+
+    private long initializeShard(int shard, boolean conservative, long initialVale) {
+        SweepShardProgressTable.Value colVal = SweepShardProgressTable.Value.of(initialVale);
+        CheckAndSetRequest initializeRequest = CheckAndSetRequest
+                .newCell(TABLE_REF, cellForShard(shard, conservative), colVal.persistValue());
+        try {
+            kvs.checkAndSet(initializeRequest);
+            return initialVale;
+        } catch (CheckAndSetException e) {
+            return getValue(getEntry(shard, conservative));
+        }
+    }
+
+    private void increaseValue(int shard, boolean conservative, long newValue) {
```

Comment:
The value stored in the database after this call might be more than `newValue`. Do we want to return the last known value, so that any consequences of this can be handled by the caller? I can't tell this by looking at this class in isolation, but it's something to bear in mind as we continue building this.